x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -500,48 +355,167 @@ private GetAliasesRequest createGetAliasesRequest(FieldCapabilitiesResponse resp
+                    }
+                    
+                    QlFieldCapability qlFieldCap = new QlFieldCapability(type, f.getType(), f.isSearchable(), f.isAggregatable());
+                    boolean isValidType = false;
+
+                    // in each list of indices, keep only those belonging to the current alias
+                    isValidType = maybeSetIndices(f.indices(), indices, qlFieldCap::setIndices) || isValidType;
+                    isValidType = maybeSetIndices(f.nonAggregatableIndices(), indices, qlFieldCap::setNonAggregatableIndices)
+                        || isValidType;
+                    isValidType = maybeSetIndices(f.nonSearchableIndices(), indices, qlFieldCap::setNonSearchableIndices) || isValidType;
+
+                    // another field type is applicable to this alias, so skip the current field type
+                    if (isEmpty(qlFieldCap.getIndices())
+                        && isEmpty(qlFieldCap.getNonAggregatableIndices())
+                        && isEmpty(qlFieldCap.getNonSearchableIndices())
+                        && (isEmpty(f.indices()) == false || isEmpty(f.nonAggregatableIndices()) == false
+                            || isEmpty(f.nonSearchableIndices()) == false)) {
+                        continue;
+                    }
+                    
+                    // a type that doesn't have any indices is valid by default, for example
+                    // "id": {"keyword": {"type": "keyword","searchable": false,"aggregatable": true}}
+                    if (isValidType || (isEmpty(f.indices()) && isEmpty(f.nonAggregatableIndices()) && isEmpty(f.nonSearchableIndices()))) {
+                        qlFieldTypes.put(type, qlFieldCap);
+                    }
+                }
+                
+                if (qlFieldTypes.size() > 0) {
+                    fields.put(fieldName, qlFieldTypes);
+                }
+            }
+
+            if (fields.size() > 0) {
+                qlFieldCaps.setFields(fields);
+                map.put(alias, qlFieldCaps);
+            }
+        }
+
+        return map;
+    }
+
+    private static boolean maybeSetIndices(String[] originalList, Set<String> virtualList, Consumer<String[]> setter) {
+        if (originalList == null) {
+            return true;
+        }
+        
+        String[] intersection = intersection(originalList, virtualList);
+        // if the indices belonging to the virtual grouping are all in the intersection no need to set the array as it's considered implicit
+        if (intersection.length == virtualList.size()) {
+            return true;
+        }
+        // if there are common indices or the original array had nothing in it, all good
+        if (intersection.length > 0 || originalList.length == 0) {
+            setter.accept(intersection);
+            return true;
+        }
+        return false;
+    }
+
+    private static String[] intersection(String[] array, Set<String> set) {
```

Comment:
Is there any value of using arrays instead of a list? Considering we have our own structures, why not use collections everywhere?

Reply:
For one, this is not a generic method to be used somewhere else. It's specific to determining if a type is valid in IndexResolver and its data types. As such, the String array always comes from ES fieldcaps output, whereas the Set is always a unique collection of indices belonging to a certain alias.